packages/lib/passport-interface/src/PassportPopup/core.ts

```diff
@@ -0,0 +1,219 @@
+
+/**
+ * An asynchronous function that opens a Zupass popup window, and returns a
+ * {@link PopupActionResult} containing either:
+ * - An encoded PCD
+ * - An encoded pending PCD
+ * - An indication that the popup was closed with no result
+ * - An indication that the popup was blocked from opening
+ *
+ * By detecting popup blockers and the popup being closed without result, this
+ * offers broader functionality than calling {@link openZupassPopup} and
+ * {@link receiveZupassPopupMessage} separately. It is added as a new API to
+ * avoid breaking backwards-compatibility for existing code.
+ */
+export async function zupassPopupExecute(
+  popupUrl: string,
+  proofUrl: string
+): Promise<PopupActionResult> {
+  const popup = await openZupassPopup(popupUrl, proofUrl);
+  // If we did not get a window from `openZupassPopup`, it was blocked
+  // See https://developer.mozilla.org/en-US/docs/Web/API/Window/open#return_value
+  // This allows the caller to gracefully handle this, e.g. by notifying the
+  // user
+  if (!popup) {
+    return { type: "popupBlocked" };
+  }
+
+  // If the popup is closed, the promise waiting for an event from the popup
+  // will not know that this has happened, and so will remain unresolved
+  // forever, leaking memory. To avoid this, we pass in an AbortController to
+  // receiveZupassPopupMessage(), which allows it to resolve.
+  // This is necessary even when using Promise.race(), as Promise.race() will
+  // not clean up the unresolved promises even after they have definitely
+  // "lost" the race.
+  const abortReceiveMessage = new AbortController();
+
+  // There's no way to receive an event indicating that a popup window has been
+  // closed, and we can't rely on the code running inside the popup to send us
+  // a message, so instead we have to check for it on a timer.
+  // We are doing this in order to detect cases in which the popup window is
+  // closed without any action being taken (e.g. a proof being generated).
+  const closePromise = new Promise<{ type: "popupClosed" }>((resolve) => {
+    const closeCheckInterval = window.setInterval(() => {
+      if (popup.closed) {
+        clearInterval(closeCheckInterval);
+        // In the case where an action is completed, the popup will send a
+        // message, and then close itself. In testing, it seems that the
+        // message is always received before we pick up on the fact that the
+        // popup is closed. However, for safety, we will avoid resolving this
+        // promise for another 250ms, to ensure that the other promise
+        // (containing the received message) definitely resolves first.
+        window.setTimeout(() => {
+          resolve({ type: "popupClosed" });
+          // In the event that the message-receiving promise has *not*
+          // resolved, we need to tell it to resolve now, to avoid a memory
+          // leak.
+          abortReceiveMessage.abort();
+        }, 250);
+      }
+    }, 100);
```

Comment:
is there a way to avoid this human-noticeable 350ms delay?

Reply:
In the previous version, there was no notification of whether or not the popup window had closed.

In this version, we can notify the caller that *either* a result was received, or that the popup was closed. In most cases, a result will be received and the above promise/timeout is irrelevant.

The code inside the popup window sends the result as a message, then closes the popup window. In my testing, the message always arrives first. However, we *really* don't want to detect the popup closing before receiving the message. The delay is intended to ensure that the message goes through first, and the `Promise.race` is won by the message promise resolving.

In the case where the window is closed and there's no message, then there would be a 0.25s delay before we conclude that there's not going to be a message, and notify the caller that the popup was closed with no result. This might be excessively safe, and maybe a much shorter delay (or none at all!) would be fine. However, since it only applies in cases where the user closed the popup window without making a proof, I think it might be OK, especially since the status quo was that the caller had no way of knowing that the popup closed at all.

Here's a video of what the delay looks like in practice:

https://github.com/proofcarryingdata/zupass/assets/20022/be3e8108-06b8-4625-b691-716a1c07c0bd